DEPS: Revert SQLAlchemy minimum version back to 1.4.36

snitish opened the pull request:
- [ ] closes #57049
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/v2.3.0.rst` file if fixing a bug or adding a new feature.

snitish committed:
BUG: Revert SQLAlchemy minimum version back to 1.4.36

snitish commented:
@mroeschke given there've been so many requests for this, thought I'd give this a try. The minimum version for SQLAlchemy was bumped from `1.4.36` to `2.0.0` about a year ago (#55524). I'm resetting it back to `1.4.36`.
My local tests ran fine with this version but is there a way to configure CI tests to run on both SQLAlchemy versions?

swarajban commented:
I don't know if just changing the versions will fix the underlying issues. When we use an old version of SA, functions like [pandasSQL_builder](https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L883) don't parse the connection object properly and return SQLite (which causes other downstream issues, eg [params](https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L2697) to not work correctly)

snitish commented:
> I don't know if just changing the versions will fix the underlying issues. When we use an old version of SA, functions like [pandasSQL_builder](https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L883) don't parse the connection object properly and return SQLite (which causes other downstream issues, eg [params](https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L2697) to not work correctly)

The reason it was incorrectly returning an SQLite object is because of this line:
https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L898

If SQLAlchemy doesn't satisfy the minimum version, the import fails and it leads to creating an SQLite object instead. If the minimum version is updated, the code works fine. Tested it on my branch after updating SQLAlchemy to 1.4.
```
>>> import pandas as pd
>>> from sqlalchemy import create_engine
f = pd.DataFrame({'a': [0, 1, 2, 3]})
df.to_sql('test', con)>>> con = create_engine('sqlite:///:memory:')
>>> df = pd.DataFrame({'a': [0, 1, 2, 3]})
>>> df.to_sql('test', con)
4
```

snitish committed:
Update pyproject.toml

swarajban commented:
> > I don't know if just changing the versions will fix the underlying issues. When we use an old version of SA, functions like [pandasSQL_builder](https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L883) don't parse the connection object properly and return SQLite (which causes other downstream issues, eg [params](https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L2697) to not work correctly)
> 
> The reason it was incorrectly returning an SQLite object is because of this line:
> 
> https://github.com/pandas-dev/pandas/blob/f46d853365b5c7b91594205730b5db3d5430bd81/pandas/io/sql.py#L898
> 
> If SQLAlchemy doesn't satisfy the minimum version, the import fails and it leads to creating an SQLite object instead. If the minimum version is updated, the code works fine. Tested it on my branch after updating SQLAlchemy to 1.4.
> 
> ```
> >>> import pandas as pd
> >>> from sqlalchemy import create_engine
> f = pd.DataFrame({'a': [0, 1, 2, 3]})
> df.to_sql('test', con)>>> con = create_engine('sqlite:///:memory:')
> >>> df = pd.DataFrame({'a': [0, 1, 2, 3]})
> >>> df.to_sql('test', con)
> 4
> ```

Ah wow, well gl! We would really appreciate this change

mroeschke commented:
> is there a way to configure CI tests to run on both SQLAlchemy versions?

There is a `Minimum Versions` build that will test with the pinned, minimum supported version of the dependencies (1.4.x), and the rest of the builds test with dependencies pinned as `>=version` so those builds should test with 2.0.x.

snitish commented:
Thanks @mroeschke . Confirmed that the minimum version tests [have passed](https://github.com/pandas-dev/pandas/actions/runs/13464255693/job/37626471840?pr=60977). The failing tests are apparently due to an unrelated issue in libsqlite-3.49.1. (See https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/16856)
I'm still not sure why these tests are passing for other PRs - they seem to be using the older 3.48.1 version - but my guess is that updating the minimum versions file may have triggered the build to re-fetch the latest versions of dependencies.

mroeschke commented:
Could you add a whatsnew note in `v2.3.0`? I think it's reasonable to backport this since many users were reporting this issue

mroeschke commented:
And after this PR, do you mind submitting one to our conda-feedstock to updating sqlalchemy version there? https://github.com/conda-forge/pandas-feedstock/blob/main/recipe/meta.yaml

snitish commented:
@mroeschke should the whatsnew note go under the 'Other Enhancements' section? Or should I create a new section for minimum version updates?

mroeschke commented:
> should the whatsnew note go under the 'Other Enhancements' section?

Yup, I think this section is appropriate

snitish committed:
Merge against master

jabbera commented:
Is this going to be merged?!?! I cannot tell you how excited this makes me. The amount of pain this issue has caused me knows no bound.

star1327p committed:
DOC Update the awkward-pandas GitHub link (#61241)

mthiboust committed:
BUG: Fix #61222: Keep index name when resampling with pyarrow dtype (#61229)

* BUG: Fix pandas-dev#61222: Keep index name when resampling with pyarrow dtype

* Update doc/source/whatsnew/v3.0.0.rst

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

j-hendricks committed:
DOC: Added docstrings to min, max, and reso (#61238)

Added docstrings to min, max, and reso

Co-authored-by: John Hendricks <jwh8zc@virginia.edu>

mroeschke committed:
STY: Bump pre-commit checks (#61246)

* Bump pre-commit version, bump clang-format and meson

* Fix type checking abbreviation

* Bump to 0.11.4

* Put minimum version at 4

* Change misc to arg-type

arthurlw committed:
ENH: Support `Series[bool]` as indexer for `iloc.__getitem__`   (#61162)

* updated indexing.py to allow iloc.__getitem__

* Updated test_iloc_mask test

* bugfix test_iloc_mask test

* bugfix test_iloc_mask

* whatsnew

* added test to test_iloc_mask

* formatting

* precommit

* added tests for series bool mask

* precommit

* reformatted tests

MartinBraquet committed:
BUG: Handle overlapping line and scatter on the same plot (#61244)

* Refactor time series plotting logic for improved clarity

Extract and streamline time series preparation steps into `prepare_ts_data`, replacing redundant logic across methods. Simplifies axis frequency handling and improves code readability while maintaining functionality.

* Add test to validate xtick alignment for scatter and line plots

This test ensures that the x-axis ticks are consistent between scatter and line plots when sharing the same axis. It addresses a potential issue related to GH#61005, verifying proper rendering of datetime x-axis labels.

* Fix bug in Series.plot misalignment for line and scatter plots

This resolves an issue where line and scatter plots were not aligned when using Series.plot. The fix ensures proper alignment and improves plot consistency. Refer to issue #61005 for further details.

* Update scatter plot test to support datetime.time data

Datetime.time is now supported in scatter plots due to added converter implementation in ScatterPlot. Removed the test expecting a TypeError and updated it to validate the new functionality.

* Refactor handling of x_data in matplotlib plotting.

Simplify and streamline the code by directly assigning x_data from the data variable and replacing the intermediate Series object with a clearer `s` variable. This improves readability and maintains the existing functionality.

* Move test_scatter_line_xticks from Series to DataFrame tests

Relocated the `test_scatter_line_xticks` test from `test_series.py` to `test_frame.py` for better alignment with DataFrame-specific functionality. This refactor ensures the test resides in the appropriate context based on its usage and focus.

* Refactor `prepare_ts_data` to improve type annotations.

Added precise type annotations to the function signature for better clarity and type checking. Replaced `data` with `series` and `kwds` with `kwargs` to enhance readability and consistency.

* Refactor test_scatter_line_xticks to simplify DataFrame creation

The DataFrame creation in the test has been streamlined for clarity and conciseness by replacing the loop with a list comprehension. This improves code readability and maintains the same functionality.

* Refactor Series import to optimize scope and maintain consistency

Moved the `Series` import inside relevant function scopes to minimize unnecessary top-level imports and align with existing import patterns. This helps improve code readability and ensures imports are only loaded where needed.

* `Reorder import statement in _make_plot method`

Moved the import of `Series` within the `_make_plot` method to comply with styling or runtime considerations. This ensures consistency and avoids potential import-related issues.

star1327p committed:
DOC: Update the last ArcticDB link in ecosystem.md (#61258)

DOC: Update ArcticDB link in ecosystem.md

mroeschke committed:
BLD: Try installing older Cython for windows free threading build (#61249)

* BLD: Try pinning ninja<1.11.1.4 for windows free threading build

* quote

* change to triple and double quotes

* Add in script instead

* Add in script instead

* Try numpy 2.2.3

* Double quotes

* Try 2.2.2

* Install older Cython

* Try commit from March 17

* try march 19 commit 0b866bf7d43ced968dba4e9726316f963aae8f3c

* Try march 18 commit b4917f731da50062f8ba53737ade7b82b4c8fcf2

* One commit after c compiler warnings PR

* Use March 20 commit 93a7d09d47d8aae0dfcea41d06f4b140a1161499

* Use cb156c48d94b7e13363ab791b16bdeeb3392f21e before vector call

* One more divmod commit

* USe divmod commit

* Use commit before divmod, undo ninja and numpy changes

DarthKitten2130 committed:
Changed term non-null to NA (#61257)

* Changed term non-null to NA

* Update pandas/io/formats/info.py

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

lesteve committed:
CI Use released numpy for Windows wheels testing (#61248)

CI Maybe fix Windows free-threaded

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

mroeschke committed:
TYP: Add ignores for numpy 2.2 updates (#61265)

* TYP: Add ignores for numpy 2.2 updates

* fix tests and plotting

* ignore pyright error

datapythonista committed:
WEB: Add pandas cookbook 3 to home page (#61279)

datapythonista committed:
WEB: Removed Coiled as a sponsor, and update past sponsors list (#61278)

jackbtlr committed:
DOC: Updated link for OVH server benchmark visualization (#61108)

datapythonista committed:
API: Rename `arg` to `func` in `Series.map` (#61264)

chilin0525 committed:
BUG: Improve ImportError message to suggest importing dependencies directly for full error details (#61084)

* Add hint to display full message for missing dependencies in pandas/init.py

* ENH: Improve import error handling to preserve original traceback

* TST: refactor testing for hard dependency package

* Update pandas/__init__.py

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

* Refactor prevent statement too long

* ENH: change unittest to verify ImportError is raised when required dependencies are missing

* TST: Use pytest.raises match parameter in test_missing_required_dependency

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

rhshadrach committed:
PERF: future_stack=True with non-MulitIndex columns (#58817)

* PERF: stack on non-MultiIndex columns

* WIP

* Use reshape instead of ravel

* arrays -> blocks

* Update test

* whatsnew

PedroM4rques committed:
BUG: OverflowError when fillna on DataFrame with a pd.Timestamp (#61208)  (#61216)

* Fix #61208: OverflowError when fillna on DataFrame with a pd.Timestamp
- Now correctly raises OutOfBoundsDatetime
- Added test_fillna_out_of_bounds_datetime()

* Comply with pre-commit and added an entry in v3.0.0.rst

* Removed flag 'inplace=True' from test and fixed the bug for this case.

robin-mader-bis committed:
BUG: Fix pyarrow categoricals not working for pivot and multiindex (#61193)

* BUG: Fix bug with DataFrame.pivot and .set_index not compatible with pyarrow dictionary categoricals

Relates to #53051

Code for fix taken and adapted from #59099

* TST: Add tests for faulty behavior relating to pyarrow categoricals

* CLN: Fix issues reported by pre-commit hooks

* TST: Fix failing tests for minimum version by ignoring obsolete deprecation warning

* DOC: Add entry for bugfix to whatsnew v3.0.0

* CLN: Refactor code and clean up according to PR feedback

* CLN: Refactor code and clean up according to PR feedback

* CLN: Refactor tests to adress PR feedback

* CLN: Refactor tests to adress PR feedback

arthurlw committed:
DOC: Add documentation for `groupby.ewm()` (#61283)

* updated doc and references

* precommit

* shortened summary

* updated according to reviewer suggestions and removed test.py

* Fixed docstring error

arthurlw committed:
DOC: Add documentation for `groupby.expanding()` (#61274)

Jeffrharr committed:
Bug: Save original index and remap after function completes (#61116)

* Save original index and remap after function completes.

* precommit passes

* use stable sorting 'mergesort' in tests

* Change sorts to `stable` instead of mergesort

* modify 'keep' to use a Literal instead of string

* address comments

* update doc to include stable sort change

rhshadrach committed:
ENH(string dtype): fallback for HDF5 with UTF-8 surrogates (#60993)

wjandrea committed:
DOC: copyedit _base.py (#61299)

No need to restate the library name

miraculixx commented:
How can we help to get this merged and included in a release? 🙏🏼

simonjayhawkins commented:
pre-commit.ci autofix

datapythonista commented:
Sorry @snitish this wasn't merged. Can you move the release note back to 3.0 and resolve the conflicts please, so we can still get this for 3.0. Thanks!

jmaupetit commented:
@snitish why closing this? Should we understand that SQLAlchemy compatibility will never be restored?
